### ISSUE TITLE
documentation: should we use build_export_depend on 'message_runtime' when building msgs?

### ISSUE THREAD

**Issue by czalidis** [ros/catkin#701]
[This file](https://github.com/ros/catkin/blob/indigo-devel/doc/howto/format2/building_msgs.rst) suggests that when building msgs, our package.xml must have:

```
<build_depend>message_generation</build_depend>
<exec_depend>message_runtime</exec_depend>
```

When running [catkin_lint](https://github.com/fkie/catkin_lint) on that package, it says that I should also add:

```
<build_export_depend>message_runtime</build_export_depend>
```

Adding `build_export_depend` sounds logical to me. Either way, the documentation needs to be updated or the [catkin_lint](https://github.com/fkie/catkin_lint) tool.

**Comment by dirk-thomas** [comment]
Both kind of dependencies `exec` as well as `build_export` will end up as a run dependency in the Debian packages. The different names are only used for semantic reasons - to help users understand what the dependency is needed for.

In case of `message_runtime` it actually serves multiple purposes: https://github.com/ros/message_runtime/blob/groovy-devel/package.xml
- the dependency on `genpy` is a clear `exec` dependency since it is required for executing the generated Python message code 
- the dependencies on the `roscpp_core` packages have two purposes
  - they are required to run executables using C++ messages since they provide shared libraries
  - the also also provide header files which are included by the generated C++ message code when code is compiled agains a the message headers

Since the `exec` usage prevails the `build_export` usage we think documenting the `exec` dependency makes more sense to the user.

Please fill a ticket against `catkin_lint`. I don't see a good reason why that tool diverges from the current `catkin` documentation.

**Comment by roehling** [comment]
I changed the logic as requested but it feels inconsistent, as 'message_runtime' is the only exec_depend while everything else listed in CATKIN_DEPENDS is a build_export_depend.

**Comment by dirk-thomas** [comment]
There is no right or wrong here. Technically `message_runtime` is both. But it is sufficient to have it as either or listed in the `package.xml`.

In the documentation it is currently referenced as an `exec` depend. Therefore I think it would be good if catkin_lint "accepts" that as valid.

But also if the user specifies it as a `build_export` dependencies or even both that should be considered "ok" by the linter.

**Comment by roehling** [comment]
Sounds good. catkin_lint will now accept both build_export_depend and exec_depend, but will specifically complain about a missing exec_depend if neither are used.

**Comment by dirk-thomas** [comment]
That is great. Since `catkin_lint` can't know which semantic meaning the missing dependency has should it recommend that "either a `build_export` or an `exec` dependency` is missing"? Otherwise it might lead the user to use a dependency type which doesn't match the semantic.

For the `message_runtime` example exec might make sense  but for other dependencies it might be the other way around (as you mentioned that most other dependencies mentioned as `CATKIN_DEPENDS` are build_export).

**Comment by roehling** [comment]
My statement was a bit unclear: the `exec_depend` logic applies merely to `message_runtime` for now. All other packages are assumed to be `build_export_depend`. As `CATKIN_DEPENDS` (like `DEPENDS`) are added to the linked libraries and include paths of dependent builds, it seems most appropriate except for corner cases like `mesage_runtime`.

However, given that there might be other corner cases in the future, maybe I should downgrade the severity to warning if there is a matching `exec_depend`.

**Comment by dirk-thomas** [comment]
It sound good the way it is now. Therefore I will mark this as closed. Thanks @roehling for the fast updates.

**Comment by roehling** [comment]
I released a new version (1.3.10) with the changes.

ping @tfoote
